### PR TITLE
module_common - handle cache directory creation collision

### PR DESCRIPTION
**Pull request by samdoran**
##### SUMMARY

Occasionally multiple workers will try to create the cache directory if it exists. This is more likely to happen when using the free strategy with a higher fork count. Catch the exception and handle it.

Closes #74895. 

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

`lib/ansible/executor/module_common.py`

### CODE REVIEW

File: lib/ansible/executor/module_common.py
Comment (nitzmahone) [inline lib/ansible/executor/module_common.py]: You shouldn't even need to try it again- the only case we want to recover from is if another worker already created it, so if we got this error, just ensure that it exists and continue on happily, otherwise re-raise the original exception.

I guess there might still be a tiny remaining race condition in that case, since `os.makedirs` is unlikely to be atomic, but adding a retry makes it more possible for workers to trip on each other, and could ultimately repeat a different incarnation of the same race for each level of the path (though *far* less likely).

If we really want to solve that potential problem, either the $controller_term_replacement_tbd needs to create the cache dir in a serialized pre-fork code path, or we'll need a dedicated lock for the creation of the tempdir. I'd probably vote for the former, since adding a new resource lock makes deadlock possible (and managing more locks is ultimately a lot more expensive than a tempdir creation).

Comment (bcoca) [reply lib/ansible/executor/module_common.py]: use `makedirs_safe()` `from ansible.utils.path`  (not as in secure but avoids errors  during concurrent access)

Comment (samdoran) [reply lib/ansible/executor/module_common.py]: > You shouldn't even need to try it again- the only case we want to recover from is if another worker already created it, so if we got this error, just ensure that it exists and continue on happily, otherwise re-raise the original exception.

I thought this looked a bit more complicated than what we discussed yesterday. I was a bit fuzzy at the end of the day and didn't remember it clearly this morning.

> I guess there might still be a tiny remaining race condition in that case, since os.makedirs is unlikely to be atomic, but adding a retry makes it more possible for workers to trip on each other, and could ultimately repeat a different incarnation of the same race for each level of the path

I was thinking as I put that together about adding a backoff wait in there. But that's adding more code and injecting waits are not something I like doing in code that needs to be performant.

> If we really want to solve that potential problem, either the $controller_term_replacement_tbd needs to create the cache dir in a serialized pre-fork code path...

I was wondering if the comment in the code about a "global function to setup" being a better place for the call to create the directory was hinting at that. I'm in favor of creating it pre-fork and avoiding the problem altogether.

Comment (nitzmahone) [reply lib/ansible/executor/module_common.py]: Yeah, that works too- just gotta find a good place that's ideally specific to the Python module subsystem that's already running pre-fork.

**Review by cloudnull — APPROVED** [review]

File: lib/ansible/executor/module_common.py
Comment (cloudnull) [inline lib/ansible/executor/module_common.py:1194]: in [py3](https://docs.python.org/3/library/os.html#os.makedirs) we can use `os.makedirs(lookup_path, exists_ok=True)`. With that we can get rid of the try/except and check condition, condensing everything here into a single line.

Comment (cloudnull) [reply lib/ansible/executor/module_common.py:1194]: I'm assuming we're only concerned with py3 given the use of `FileExistsError`; that didn't exist in py2. If we need py2 support we'd need something like so to trap the exception in a py2 compatible way and the above comment wont work; the `exists_ok` flag was new in py3.

``` python
import errno

class FileExistsError(OSError):
    def __init__(self, message):
        super(FileExistsError, self).__init__(errno.EEXIST, message)
```

**Review by cloudnull — CHANGES_REQUESTED** [review]

File: lib/ansible/executor/module_common.py
Comment (bcoca) [reply lib/ansible/executor/module_common.py:1194]: @cloudnull see `makedirs_safe` for older python compatibility (probably will be in backport), but since we require python >=3.8 on controller, `exist_ok=True` should be good enough

Comment (samdoran) [reply lib/ansible/executor/module_common.py:1194]: For changes in `devel`, we are only concerned with supporting Python >3.8. If we decide to backport an equivalent fix, we could use `makedirs_safe()`.

Comment (nitzmahone) [reply lib/ansible/executor/module_common.py:1194]: Since we're intending to backport this fix, it'll have to remain a `try/except OSError:` with a double-check in the exception handler. Once it's landed in devel and our CI is able to handle py3-only controller tests (which is still WIP), we can update to use py3-only bits.

**Comment by samdoran** [comment]
The latest changes in this PR will required split controller/remote before tests will pass. I'm not sure if we want to wait for that or maybe take another approach to fixing it.

File: lib/ansible/executor/module_common.py
Comment (bcoca) [reply lib/ansible/executor/module_common.py:1194]: @nitzmahone no need, makedirs_safe already does that

File: lib/ansible/executor/module_common.py
Comment (cloudnull) [inline lib/ansible/executor/module_common.py:1190]: there's no need to check if the path exists with the following line.

File: lib/ansible/executor/module_common.py
Comment (nitzmahone) [reply lib/ansible/executor/module_common.py:1194]: It involves `unfrackpath` among other things ... :fearful:

File: lib/ansible/executor/module_common.py
Comment (nitzmahone) [reply lib/ansible/executor/module_common.py:1190]: Right, but we can't merge or backport that right now.

File: lib/ansible/executor/module_common.py
Comment (samdoran) [reply lib/ansible/executor/module_common.py:1194]: That was my hesitancy as well. I think I'll go back to the plain vanilla `try/except`.

File: lib/ansible/executor/module_common.py
Comment (cloudnull) [reply lib/ansible/executor/module_common.py:1190]: Assuming py3 is the only thing being supported, why wouldn't we be able to backport that? Given https://github.com/python/cpython/blob/main/Lib/os.py#L213, the method is running the same check?

Comment (cloudnull) [reply lib/ansible/executor/module_common.py:1190]: more for my curiosity than anything else?

Comment (samdoran) [reply lib/ansible/executor/module_common.py:1190]: The Python >=3.8 requirement is only for Ansible >= 2.12 (current `devel`). Previous version still support Python 2.

Comment (samdoran) [reply lib/ansible/executor/module_common.py:1190]: We can't merge this change currently because in our CI, the Ansible control node and the remote node are the same. For systems that only have Python 2 (such as CentOS 6) or older versions of Python 3, tests will fail (which is what is happening currently).

Comment (bcoca) [reply lib/ansible/executor/module_common.py:1190]: @samdoran for now you could just point it at makedirs_safe() and when controller spilt happens, i'll update in  #75108 with the rest of makedirs_safe usage

Comment (samdoran) [reply lib/ansible/executor/module_common.py:1190]: I added a note that we can switch to this method once split controller/remote is merged.

File: lib/ansible/executor/module_common.py
Comment (samdoran) [inline lib/ansible/executor/module_common.py:37]: Unused import.

File: lib/ansible/executor/module_common.py
Comment (samdoran) [inline lib/ansible/executor/module_common.py:1118]: Unused.

**Review by cloudnull — APPROVED** [review]

File: lib/ansible/executor/module_common.py
Comment (nitzmahone) [reply lib/ansible/executor/module_common.py:1190]: @cloudnull for more detail: the `ansible-test` changes that support split testing on the controller and remotes is landing in devel Real Soon Now, but until it does, we can't merge anything that doesn't work back to Python 2.6 without breaking all testing on those targets or disabling all module tests on < py3.8. It's a BIG change to the internals of the way all that stuff works, and breaks a lot of assumptions that module tests often make about the test target host being the same as the controller host. Since lots of external collections rely on `ansible-test`, we're taking time to maximize backwards compatibility to keep their tests working, rather than just slamming in the changes we need and letting the community pick up the pieces (otherwise it would've landed long ago in 2.11).

**Review by nitzmahone — APPROVED** [review]

**Pull request merged**
